filters/auth/grant.go

```diff
@@ -126,29 +126,43 @@ func (f *grantFilter) refreshTokenIfRequired(c *cookie, ctx filters.FilterContex
 	}
 }
 
-func (f *grantFilter) setAccessTokenHeader(req *http.Request, token string) {
+func (f *grantFilter) setupToken(token *oauth2.Token, tokeninfo map[string]interface{}, ctx filters.FilterContext) error {
 	if f.config.AccessTokenHeaderName != "" {
-		req.Header.Set(f.config.AccessTokenHeaderName, authHeaderPrefix+token)
+		ctx.Request().Header.Set(f.config.AccessTokenHeaderName, authHeaderPrefix+token.AccessToken)
 	}
-}
 
-func (f *grantFilter) createTokenContainer(token *oauth2.Token, tokeninfo map[string]interface{}) (tokenContainer, error) {
 	subject := ""
 	if f.config.TokeninfoSubjectKey != "" {
 		if s, ok := tokeninfo[f.config.TokeninfoSubjectKey].(string); ok {
 			subject = s
 		} else {
-			return tokenContainer{}, fmt.Errorf("tokeninfo subject key '%s' is missing", f.config.TokeninfoSubjectKey)
+			return fmt.Errorf("tokeninfo subject key '%s' is missing", f.config.TokeninfoSubjectKey)
 		}
 	}
 
 	tokeninfo["sub"] = subject
 
-	return tokenContainer{
+	if len(f.config.GrantTokeninfoKeys) > 0 {
+		for key := range tokeninfo {
+			if !contains(f.config.GrantTokeninfoKeys, key) {
+				delete(tokeninfo, key)
+			}
+		}
+	}
+
+	// By piggy-backing on the OIDC token container,
+	// we gain downstream compatibility with the oidcClaimsQuery filter.
+	ctx.StateBag()[oidcClaimsCacheKey] = tokenContainer{
 		OAuth2Token: token,
 		Subject:     subject,
 		Claims:      tokeninfo,
-	}, nil
+	}
```

Comment:
@szuecs 
I just moved this piece around without changing. Token container is stored in statebag to make https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#grantclaimsquery work which is an alias for oidcClaimsQuery. I agree it was not the best idea but it slipped through the review back then.

I can fix it in the followup PR - we'd need to make `oidcClaimsQuery` parameterized on key value. It could also break some configurations if users use `oidcClaimsQuery` instead of `grantClaimsQuery` after `oauthGrant`.